src/postProposalVersions.ts

```diff
@@ -0,0 +1,327 @@
+// Takes a CSV file, creates JSON bodies for POST /proposalVersions, posts them.
+import { readFileSync } from 'fs';
+import { AssertionError } from 'assert';
+import { parse as csvParse } from 'csv-parse/sync';
+import { parse as argParse } from 'ts-command-line-args';
+import axios, { AxiosError } from 'axios';
+
+interface Args {
+  inputFile: string;
+  applicantColumnName: string;
+  applicationFormId: number;
+  proposalExternalIdColumnName: string;
+  bearerToken: string;
+  apiUrl: string;
+}
+
+type CsvRow = Record<string, string>;
+
+function assertIsCsvRow(row: unknown): asserts row is CsvRow {
+  if (row === null
+    || typeof row !== 'object'
+    || Object.keys(row).length === 0
+    || Object.values(row).length === 0) {
```

Comment:
I kind of alluded to this during our call, but I hadn't fully thought this type guard through at that time: I think we want to be a little bit stricter in our checks here, because we're telling TypeScript some things we're actually not quite sure of. In particular, we want to make sure all the keys and values are actually `string`s, which we're not checking. And, on the other hand, as [`null` is a distinct type and not a universal subtype](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#null-and-undefined), we don't need to both check for `null`ness and check the `typeof` - the latter covers both. Likewise, the API of `Object` is that every key has a value, and vice versa; we don't have to check that there are a nonzero number of both.

```suggestion
type CsvRow = Record<string, string>;

function assertIsCsvRow(row: unknown): asserts row is CsvRow {
  if (typeof row !== 'object'
    || Object.keys(row).length === 0
    || Object.keys(row).some((key) => typeof key !== 'string')
    || Object.values(row).some((value) => typeof value !== 'string')) {
```

Reply:
The reason to check if the value is `null` first is that `typeof null` is `object`. The rest seems good to me.

Reply:
Yes, quite right - I was focused too strongly on the TypeScript discussion we were having earlier and forgot about the JavaScript side. Good catch!